doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+
+* A question mark indicates the preceding symbol or group may be omitted, that is, it can appear at most one time.
+
+
+## Character set
+<a id="character-set"/>
+
+Arblang source is represented as a sequence of Unicode characters in normalization form C (NFC), excluding unassigned characters, surrogate code points, non-characters, and byte order marks. It is the responsibility of the interpreting environment to transform any particular source encoding to this representation.
+
+
+## Tokenization
+<a id="tokenization"/>
+
+A lexically valid Arblang source must admit a tokenization as a sequence of ***comment*** tokens, ***whitespace*** tokens, ***symbol*** tokens, ***string-literal*** tokens, ***superscript-literal*** tokens, ***numeric-literal*** tokens, and punctuation tokens.
+
+A token has an associated value, which may be empty. If not empty, it is the result of applying a token-specific canonicalization process.
+With the exception of string literals, canonicalization first applies Unicode compatibility normalization NFKC to the token text, then further processing as detailed below.
+
+
+### Comments
+<a id="comments"/>
+
+A comment is introduced by `#` and then extends to the end of the line. It is the only context in which a line end is not regarded simply as whitespace.
+
+> ***comment*** ::= U+0023 `#` NUMBER SIGN _comment-character_* _new-line_
+>
+> _comment_character_ ::= [^\\u{000A}\\u{000B}\\u{000C}\\u{000D}\\u{0085}\\u{2028}\\u{2029}]
+>
+> _new-line_ ::= _cr_ | _lf_ | _cr_ _lf_ | _other-newline_
+>
+> _lf_ ::= U+000A LINE FEED
+>
+> _cr_ ::= U+000D CARRIAGE RETURN
+>
+> _other-new-line_ ::= U+0085 NEXT LINE | U+2028 LINE SEPARATOR | U+2029 PARAGRAPH SEPARATOR
+
+#### Token value
+
+A ***comment*** token has no value.
+
+#### Notes
+
+A _comment-character_ is any character that is not the first character of _new-line_. New line characters and sequences correspond to the [line boundaries requirement RL1.6 of Unicode Regular Expressions](https://www.unicode.org/reports/tr18/#RL1.6).
+
+
+### Whitespace
+<a id="whitespace"/>
+
+In some expressions, where it denotes multiplication, whitespace is mandatory, but otherwise whitespace is not significant. It comprises any non-empty sequence of characters with the White_Space property.
+
+> ***whitespace*** ::= {White_Space}\+
+
+#### Token value
+
+A ***whitespace*** token has no value.
+
+
+### Symbols
+<a id="symbols"/>
+
+Symbols correspond to Arblang identifiers and keywords. Arblang keywords are generally contextual: the interpretation of a symbol as an identifier or a keyword is determined in parsing.
+
+> ***symbol*** ::= _symbol-start_ _symbol-continue_\* | _symbol-degree-prefixed_
+>
+> _symbol-start_ ::= {L} | {Nl} | {Other_ID_Start}
+>
+> _symbol-continue_ ::= _symbol-start_ | {Nd} | {Mn} | {Mc} | {Pc} | _prime-mark_
+>
+> _prime-mark_ ::= U+0027 `'` APOSTROPHE | U+02B9 `ʹ` MODIFIER LETTER PRIME | U+2032 `′` PRIME
```

Comment:
These are potentially confusing since hard to separate visually. Maybe better to decide on one?

Reply:
They are both mapped to the same character in the token value, so that `a′` and `a'` are just different ways of writing the same identifier.

Reply:
I think that would make it worse, since they look differently in source (depending on font etc), but map to the same internal name. It's like we translate `phi` to a unicode symbol but also allow a literal symbol phi.

Reply:
We can keep apostrophe and lose the prime mark, but I honestly don't think this is going to be problematic. (Certainly to allow both and _not_ conflate them would sow confusion.)